weave-js/src/components/PagePanelComponents/Home/Browse3/pages/PlaygroundPage/PlaygroundChat/useChatCompletionFunctions.tsx

```diff
@@ -162,7 +392,140 @@ export const useChatCompletionFunctions = (
     }
   };
 
-  return {handleRetry, handleSend};
+  /*
+   * Streaming-aware retry. Performs same state modifications as handleRetry
+   * but uses the streaming completions endpoint so the UI can progressively
+   * show the regenerated answer.
+   */
+  const handleRetryStream = async (
+    callIndex: number,
+    messageIndex: number,
+    choiceIndex?: number
+  ) => {
+    try {
+      setPlaygroundStateField(callIndex, 'loading', true);
+
+      const updatedStates = filterNullMessagesFromStates(
+        playgroundStates.map((state, index) => {
+          if (index === callIndex) {
+            if (choiceIndex !== undefined) {
+              return appendChoiceToMessages(state, choiceIndex);
+            }
+            const updatedState = JSON.parse(JSON.stringify(state));
+            if (updatedState.traceCall?.inputs?.messages) {
+              updatedState.traceCall.inputs.messages =
+                updatedState.traceCall.inputs.messages.slice(
+                  0,
+                  messageIndex + 1
+                );
+            }
+            return updatedState;
+          }
+          return state;
+        })
+      );
+
+      // Accumulate chunks into a single response object while streaming
+      let aggregatedRes: any = createAggregatedResponse();
+
+      // Kick off streaming request
+      const res = await makeCompletionStreamRequest(
+        callIndex,
+        updatedStates,
+        chunk => {
+          const delta = chunk.choices?.[0]?.delta;
+          if (delta?.content) {
+            aggregatedRes.choices[0].message.content += delta.content;
+            // Live update UI
+            setPlaygroundStates(prev => {
+              const next = [...prev];
+              const tgt = next[callIndex];
+              // Preserve existing tool calls when updating state
+              const existingToolCalls =
+                (tgt.traceCall?.output as any)?.choices?.[0]?.message
+                  ?.tool_calls || [];
+              const currentToolCalls =
+                aggregatedRes.choices[0].message.tool_calls || [];
+              tgt.traceCall = {
+                ...tgt.traceCall,
+                output: {
+                  ...aggregatedRes,
+                  choices: [
+                    {
+                      ...aggregatedRes.choices[0],
+                      message: {
+                        ...aggregatedRes.choices[0].message,
+                        tool_calls:
+                          currentToolCalls.length > 0
+                            ? currentToolCalls
+                            : existingToolCalls,
+                      },
+                    },
+                  ],
+                },
+              } as any;
```

Comment:
I realize the type might be too nested/complicated to define exhaustively. If that's the case, maybe we could at least type the skeleton and push `any`s down a few levels.